native-image argument files break when arguments contain spaces

This is graal master built from source (revision 12999967bd12d81440b4daa3a7d5101c3424131f). After the patch which adds support for argument files for the native image generator runner's JVM invocation (rev 7c4a040b2a2b7fbe7ea29cf4c76ce429ea24bc68), passing JVM options containing spaces breaks. For example passing a java property that contains a space, but is properly quoted on the command line. That worked before 7c4a040b2a2b7fbe7ea29cf4c76ce429ea24bc68 and breaks after.

Steps to reproduce:
```
pushd substratevm
cat > HelloWorld.java <<EOF
public class HelloWorld {
  public static void main(String[] args) {
    System.out.println("Hello World!");
  }
}
EOF
javac HelloWorld.java
mx --native=native-image --components="Native Image" build
popd
# Then using the result
native-image -Dorg.graalvm.config="foo bar" -cp substratevm HelloWorld hello           <--- Breaks
native-image -cp substratevm HelloWorld hello                                          <--- Works
```

Build JDK in use doesn't matter. I can reproduce with JDK 11, JDK 16 and 17.

Actual:
```
$ native-image -Dorg.graalvm.config="foo bar" -cp substratevm HelloWorld hello
Error occurred during initialization of VM
java.lang.Error: com.oracle.svm.hosted.NativeImageSystemClassLoader
	at java.lang.ClassLoader.initSystemClassLoader(java.base@16.0.2/ClassLoader.java:1990)
	at java.lang.System.initPhase3(java.base@16.0.2/System.java:2132)
Caused by: java.lang.ClassNotFoundException: com.oracle.svm.hosted.NativeImageSystemClassLoader
	at jdk.internal.loader.BuiltinClassLoader.loadClass(java.base@16.0.2/BuiltinClassLoader.java:636)
	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(java.base@16.0.2/ClassLoaders.java:182)
	at java.lang.ClassLoader.loadClass(java.base@16.0.2/ClassLoader.java:519)
	at java.lang.Class.forName0(java.base@16.0.2/Native Method)
	at java.lang.Class.forName(java.base@16.0.2/Class.java:466)
	at java.lang.ClassLoader.initSystemClassLoader(java.base@16.0.2/ClassLoader.java:1976)
	at java.lang.System.initPhase3(java.base@16.0.2/System.java:2132)

Error: Image build request failed with exit status 1
```

Expected: native image generation works. Like this:
```
$ native-image  -cp-Dorg.graalvm.config="foo bar" substratevm HelloWorld hello
[hello:178204]    classlist:   1,298.60 ms,  0.96 GB
[hello:178204]        (cap):     535.22 ms,  0.96 GB
[hello:178204]        setup:   2,141.86 ms,  0.96 GB
[hello:178204]     (clinit):     217.02 ms,  1.72 GB
[hello:178204]   (typeflow):   3,546.16 ms,  1.72 GB
[hello:178204]    (objects):   3,630.22 ms,  1.72 GB
[hello:178204]   (features):     610.76 ms,  1.72 GB
[hello:178204]     analysis:   8,261.82 ms,  1.72 GB
[hello:178204]     universe:     931.04 ms,  1.72 GB
[hello:178204]      (parse):     815.59 ms,  1.76 GB
[hello:178204]     (inline):   1,071.56 ms,  2.31 GB
[hello:178204]    (compile):   8,995.11 ms,  3.19 GB
[hello:178204]      compile:  11,450.09 ms,  3.19 GB
[hello:178204]        image:   1,004.56 ms,  3.19 GB
[hello:178204]        write:     231.79 ms,  3.19 GB
[hello:178204]      [total]:  25,577.91 ms,  3.19 GB
# Printing build artifacts to: /disk/graal/upstream-sources/graalvm/hello.build_artifacts.txt
./hello
Hello World!
```

The root cause of this is that this new \@argfile support relies on OpenJDK's \@argfile support. So in a way the reproducer for plain java would be:

```
cat > myargfile.txt <<EOF
-Dmyprop=foo bar
-XshowSettings:properties
-version
EOF
java \@myargfile.txt
Error: Could not find or load main class bar
Caused by: java.lang.ClassNotFoundException: bar
```

Quoting the arguments in \@argfile(s) fixes that issue.

--- a/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java
+++ b/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java
@@ -50,6 +50,7 @@ import java.util.ListIterator;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Properties;
+import java.util.StringJoiner;
 import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
@@ -1301,7 +1302,22 @@ public class NativeImage {
     protected static String createVMInvocationArgumentFile(List<String> arguments) {
         try {
             Path argsFile = Files.createTempFile("vminvocation", ".args");
-            String joinedOptions = String.join("\n", arguments);
+            StringJoiner joiner = new StringJoiner("\n");
+            for (String arg : arguments) {
+                // Options in @argfile need to be properly quoted as
+                // this relies on the JDK's @argfile parsing when the
+                // native image generator is being launched.
+                String quoted = SubstrateUtil.quoteShellArg(arg);
+                // @argfile rules for Windows quirk: backslashes don't need to be
+                // escaped if the option they are used in isn't quoted. If it is
+                // though, then they need to be escaped. This might mean that
+                // user-supplied arguments containing '\' will be double escaped.
+                if (quoted.startsWith("'")) {
+                    quoted = quoted.replace("\\", "\\\\");
+                }
+                joiner.add(quoted);
+            }
+            String joinedOptions = joiner.toString();
             Files.write(argsFile, joinedOptions.getBytes());
             argsFile.toFile().deleteOnExit();
             return "@" + argsFile;

